app/Jobs/CloneImagesOrVideos.php

```diff
@@ -0,0 +1,480 @@
+     **/
+    public bool $cloneAnnotations;
+
+    /**
+     * Array containing annotation label ids.
+     *
+     * @var array
+     **/
+    public array $onlyAnnotationLabels;
+
+    /**
+     * Boolean for cloning file labels.
+     *
+     * @var bool
+     **/
+    public bool $cloneFileLabels;
+
+    /**
+     * Array containing file label ids.
+     *
+     * @var array
+     **/
+    public array $onlyFileLabels;
+
+
+    /**
+     * Create a new job instance.
+     *
+     * @param CloneVolume $request containing the project, volume, new volume name and ids of files and labels.
+     *
+     * @return void
+     */
+    public function __construct($request)
+    {
+        $this->project = $request->project;
+        $this->volume = $request->volume;
+        $this->cloneName = $request->input('name', $this->volume->name);
+        $this->onlyFiles = $request->input('only_files', []);
+        $this->cloneAnnotations = $request->input('clone_annotations', false);
+        $this->onlyAnnotationLabels = $request->input('only_annotation_labels', []);
+        $this->cloneFileLabels = $request->input('clone_file_labels', false);
+        $this->onlyFileLabels = $request->input('only_file_labels', []);
+
+    }
+
+    public function handle()
+    {
+        DB::transaction(function () {
+            $project = $this->project;
+            $volume = $this->volume;
+            $onlyFiles = $this->onlyFiles;
+            $cloneAnnotations = $this->cloneAnnotations;
+            $onlyAnnotationLabels = $this->onlyAnnotationLabels;
+            $cloneFileLabels = $this->cloneFileLabels;
+            $onlyFileLabels = $this->onlyFileLabels;
+
+            $copy = $volume->replicate();
+            $copy->name = $this->cloneName;
+            $copy->save();
+
+            if ($volume->isImageVolume()) {
+                $this->copyImages($volume, $copy, $onlyFiles);
+                if ($cloneAnnotations) {
+                    $this->copyImageAnnotation($volume, $copy, $onlyFiles, $onlyAnnotationLabels);
+                }
+                if ($cloneFileLabels) {
+                    $this->copyImageLabels($volume, $copy, $onlyFiles, $onlyFileLabels);
+                }
+            } else {
+                $this->copyVideos($volume, $copy, $onlyFiles);
+                if ($cloneAnnotations) {
+                    $this->copyVideoAnnotation($volume, $copy, $onlyFiles, $onlyAnnotationLabels);
+                }
+                if ($cloneFileLabels) {
+                    $this->copyVideoLabels($volume, $copy, $onlyFiles, $onlyFileLabels);
+                }
+            }
+            if ($copy->files()->exists()) {
+                ProcessNewVolumeFiles::dispatch($copy);
+            }
+
+            //save ifdo-file if exist
+            if ($volume->hasIfdo()) {
+                $this->copyIfdoFile($volume->id, $copy->id);
+            }
+
+            $project->addVolumeId($copy->id);
+
+            $copy->flushThumbnailCache();
```

Comment:
Why is this needed?

Reply:
If I delete addVolmeId(), my tests fail with "Error: Call to a member function videos()/images() on null".
But the flushThumbnailCache method was unnecessary, so I deleted it.

Reply:
If I delete addVolumeId(), my tests fail with "Error: Call to a member function videos() on null".
But the last line is unecessary. I will remove it.

Reply:
Yes, I was just referring to `flushThumbnailCache`. Thanks!